File: solutions/cpp/combinationSum/solution_test.cpp
//
// Created by mingyi on 26.05.20.
//

#include "gtest/gtest.h"
#include "gmock/gmock.h"

#include "solution.cpp"
#include "../include/structures.hpp"

using namespace std;
using namespace testing;

typedef TestStruct<tuple<vector<int>, int>, vector<vector<int>>> TStruct;

class CombinationSumTest : public ::testing::TestWithParam<TStruct> {};

TEST_P(CombinationSumTest, Test) {
  TStruct tt = GetParam();
  auto[candidates, target] = tt.input;
  vector<vector<int>> actual = combinationSum(candidates, target);
  ASSERT_THAT(actual, UnorderedElementsAreArray(tt.expected));
}

INSTANTIATE_TEST_SUITE_P(Tests, CombinationSumTest, ::testing::Values(
    TStruct{
        .input = make_tuple(vector<int>{2, 3, 6, 7}, 7),
        .expected = {{7}, {2, 2, 3}}
    },
    TStruct{
        .input = make_tuple(vector<int>{2, 3, 5}, 8),
        .expected = {{2, 2, 2, 2}, {2, 3, 3}, {3, 5}}
    }
));

File: solutions/golang/combinationSum/solution_test.go
package main

import (
  "testing"

  "github.com/stretchr/testify/assert"
)

func TestCombinationSum(t *testing.T) {
  tests := []struct {
    candidates []int
    target     int
    expected   [][]int
  }{
    {
      candidates: []int{2, 3, 6, 7},
      target:     7,
      expected:   [][]int{{2, 2, 3}, {7}},
    }, {
      candidates: []int{2, 3, 5},
      target:     8,
      expected:   [][]int{{2, 2, 2, 2}, {2, 3, 3}, {3, 5}},
    },
  }
  for _, tt := range tests {
    actual := combinationSum(tt.candidates, tt.target)
    assert.ElementsMatchf(t, actual, tt.expected, "got %v, expected %v", actual, tt.expected)
  }
}
